### PR TITLE
PodTemplate for Workload

### LINKED ISSUE #1004: Compose a Workload object using PodTemplate objects
**What would you like to be added**:

A field in the `PodSet` struct that references a Pod or [PodTemplate object](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-template-v1/)

```golang
type PodSet struct {
  Name string
  Count int32
  Template *corev1.PodTemplateSpec // keep for backwards compatibility
  PodRef *TemplateReference
  PodTemplateRef *TemplateReference
}
```

Only one of (Template, PodRef or PodTemplateRef) can be set.

Caveats:
- A Workload wouldn't be ready for admission until Kueue has seen all the referenced PodTemplate objects.
- The PodTemplate objects should have a label that allows to:
    - reference the Job/MPIJob/etc object
    - filter from a webhook so that we can implement immutability checks.
 - The Job/MPIJob/etc should be the owner of the PodTemplate object so that it can be deleted in cascade when the job is deleted.

**Why is this needed**:

1. We can rely on the apiserver validation for the PodTemplate object.
2. Avoid duplication of the template on a ProvisioningRequest https://github.com/kubernetes/autoscaler/pull/5848 
3. Allow the Workload object to host more specs without worrying about etcd size limits
4. For #976, it feels overkill to have to duplicate the entire podspec just to represent one Pod

If JobSet were to implement a similar approach, we would also reuse the same PodTemplate objects created by users.

**Completion requirements**:

This enhancement requires the following artifacts:

- [x] Design doc
- [x] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by stuton**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Ref: #1004

#### Which issue(s) this PR fixes:

Fixes #1004 

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by stuton** [comment]
Duplicated PR #1052 was closed. This is a new one

File: apis/kueue/v1beta1/workload_types.go [hunk 1]
```diff
@@ -131,6 +131,9 @@ type PodSet struct {
 	//
 	// +optional
 	MinCount *int32 `json:"minCount,omitempty"`
+
+	// PodTemplateName is the name of the PodTemplate the Workload is associated with.
+	PodTemplateName *string `json:"podTemplateName,omitempty"`
```
Comment (trasc) [inline apis/kueue/v1beta1/workload_types.go:136]:
```suggestion

	// PodTemplateName is the name of the PodTemplate the Workload is associated with.
	//
	// +optional
	PodTemplateName *string `json:"podTemplateName,omitempty"`
```

File: apis/kueue/v1beta1/workload_types.go [hunk 2]
```diff
@@ -113,7 +113,7 @@ type PodSet struct {
 	// the keys in the nodeLabels from the ResourceFlavors considered for this
 	// Workload are used to filter the ResourceFlavors that can be assigned to
 	// this podSet.
-	Template corev1.PodTemplateSpec `json:"template"`
+	Template *corev1.PodTemplateSpec `json:"template"`
```
Comment (trasc) [inline apis/kueue/v1beta1/workload_types.go:116]:
```suggestion
	//
	// +optional
	Template *corev1.PodTemplateSpec `json:"template"`
```

File: pkg/controller/core/workload_controller.go [hunk 3]
```diff
@@ -134,6 +135,20 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+			continue
+		}
+		var pt corev1.PodTemplate
+		if err := r.client.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("%s-%s", wl.Name, ps.Name), Namespace: wl.Namespace}, &pt); err != nil {
+			continue
+		}
+		wl.Spec.PodSets[i].PodTemplateName = &pt.Name
+		if err := r.client.Update(ctx, &wl); err != nil {
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
```
Comment (trasc) [inline pkg/controller/core/workload_controller.go:149]: I think we should  not push the template copy to the API, and only use the fully populated wl internally (pass it to cache , queue etc.)

**Pull request closed without merging**